text/0001-single-page-applications.md

```diff
@@ -0,0 +1,34 @@
+# Single Page Applications
+- Start Date: 04/23/2019
+- RFC PR: https://github.com/openmrs/openmrs-rfc-frontend/pull/1
+
+## Decision, including impact on distributions
+Core OpenMRS frontend modules will be implemented within a master single page application. Distributions are encouraged
```

Comment:
I guess this takes care of OpenMRS Web Apps. Correct?

https://wiki.openmrs.org/x/C4KIBQ

https://wiki.openmrs.org/x/_o2QBQ

Reply:
Oo thanks for linking me to those -- I hadn't seen them and didn't know about packaged web apps in OpenMRS. Is this how distributions like Bahmni and Ampath's distro are writing single page applications right now? Creating their own packaged web app?

It looks like packaged web apps have their own HTML file -- my proposal in this PR is to consolidate all HTML files into one master single page application. Modules that plug into that master single page application can have javascript, css, img, and font assets. But they would not be able to have their own HTML file.

Thoughts on that? ^

Reply:
That looks fine to me. Bahmni and Ampath are not Open Web Apps. I just mentioned it for a few Open Web Apps that are already in existence. Though not as many as the JSP, GSP ones.

Reply:
Is the idea that the *all* distributions will include a giant SPA that includes *all* the the front ends built by all contributing distributions, or will there be a "pick-and-choose" model where each distribution will build and compile their own SPA based on the modules they want to use?

Reply:
Update... sounds like reading further comment the answer is the latter, the "pick-and-choose" model...we should probably make that clear in this doc.

Reply:
Yeah, the pick and choose model. Some of the technical details are not fully fleshed out, but a lot of it is. We are proposing to use [modules](https://github.com/openmrs/openmrs-rfc-frontend/pull/2), [import maps](https://github.com/WICG/import-maps), and [single spa applications](https://single-spa.js.org/docs/building-applications.html) to structure the code. Distributions will be able to author their own single spa applications that are modules that get put into the import map.

Some core modules will be baked into the master single page application, but after that it will be up to the distributions to add their own modules.

Reply:
@jonathandick my plan is to have a separate RFC proposal that goes over single-spa applications. You are right that it's a top-level router -- another way of thinking about it is that single-spa will provide structure to the modules in our SPA.